protocol/morse/store.go

```diff
@@ -0,0 +1,165 @@
+	defaultCacheCleanupInterval = 10 * time.Minute
+)
+
+// sessionSanctionKey creates a unique key for session-based sanctions
+// Format: app_addr:session_key:endpoint_addr
+func sessionSanctionKey(appAddress, sessionKey, endpointAddr string) string {
+	return fmt.Sprintf("%s:%s:%s", appAddress, sessionKey, endpointAddr)
+}
+
+// EndpointStore maintains sanctions for Morse endpoints
+// It provides both permanent sanctions and time-limited session-based sanctions
+// with automatic expiration through go-cache.
+type EndpointStore struct {
+	logger polylog.Logger
+
+	// Mutex for synchronized access to the permanent sanctions map
+	mu sync.RWMutex
+
+	// permanentSanctions stores endpoints with permanent sanctions
+	// These sanctions never expire and will require manual intervention to remove
+	permanentSanctions map[protocol.EndpointAddr]sanction
+
+	// sessionSanctions stores endpoints with session-limited sanctions
+	// The key is a combination of app address, session key, and endpoint address
+	// These sanctions automatically expire after defaultSessionSanctionExpiration
+	sessionSanctions *cache.Cache
+}
+
+// NewEndpointStore creates a new EndpointStore
+func NewEndpointStore(logger polylog.Logger) *EndpointStore {
+	return &EndpointStore{
+		logger:             logger,
+		permanentSanctions: make(map[protocol.EndpointAddr]sanction),
+		sessionSanctions:   cache.New(defaultSessionSanctionExpiration, defaultCacheCleanupInterval),
+	}
+}
+
+// AddSanction adds a sanction for an endpoint based on observed behavior
+// Two types of sanctions can be added:
+//  1. Permanent sanctions: These never expire and require manual intervention to remove.
+//     Used for serious errors like validation failures that may indicate malicious behavior.
+//  2. Session-based sanctions: These automatically expire after a set duration (1 hour by default).
+//     Used for temporary issues like timeouts or connection problems.
+func (es *EndpointStore) AddSanction(
+	endpointAddr protocol.EndpointAddr,
+	appAddress string,
+	sessionKey string,
+	errorType protocolobservations.MorseEndpointErrorType,
+	sanctionType protocolobservations.MorseSanctionType,
+	reason string,
+	sessionChain string,
+	sessionHeight int,
+) {
+	// Create a sanction record with detailed information
+	newSanction := sanction{
+		Type:          sanctionType,
+		Reason:        reason,
+		ErrorType:     errorType,
+		CreatedAt:     time.Now(),
+		SessionChain:  sessionChain,
+		SessionHeight: sessionHeight,
+	}
+
+	logger := es.logger.With(
+		"endpoint", string(endpointAddr),
+		"app_addr", appAddress,
+		"reason", reason,
+		"session_chain", sessionChain,
+		"session_height", sessionHeight,
+		"session_key", sessionKey,
+	)
+
+	switch sanctionType {
+	case protocolobservations.MorseSanctionType_MORSE_SANCTION_PERMANENT:
+		// Permanent sanctions are stored in a map and never expire
+		// They require manual intervention to remove
```

Comment:
Can we update the comment on what the manual process is for removing these?

Reply:
We're repeating this comment many times. I think we can delete it in this specific location and update above.